### PR TITLE
En 4126 headers interceptors refactor

### PR DESCRIPTION
**Pull request by iulianpascalau**
Headers interceptors refactoring

### CODE REVIEW

File: process/block/interceptedBlocks/interceptedBlockHeader.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	inHdr.processFields(arg.HdrBuff)
+
+	return inHdr, nil
+}
+
+func (inHdr *InterceptedHeader) processFields(txBuff []byte) {
+	inHdr.hash = inHdr.hasher.Compute(string(txBuff))
+
+	isHeaderForCurrentShard := inHdr.shardCoordinator.SelfId() == inHdr.HeaderHandler().GetShardID()
+	isMetachainShardCoordinator := inHdr.shardCoordinator.SelfId() == sharding.MetachainShardId
```
Comment (sasurobert) [inline process/block/interceptedBlocks/interceptedBlockHeader.go]: this is constant for a complete epoch. GetShardID could return metachainShardID and would be ok.

File: process/block/interceptedBlocks/interceptedBlockHeader_test.go [hunk 2]
```diff
@@ -0,0 +1,349 @@
+package interceptedBlocks_test
+
+import (
+	"testing"
+
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	block2 "github.com/ElrondNetwork/elrond-go/data/block"
```
Comment (sasurobert) [inline process/block/interceptedBlocks/interceptedBlockHeader_test.go]: change from block2

File: process/block/interceptedBlocks/interceptedBlockHeader_test.go [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+		Signature:        []byte("signature"),
+		MiniBlockHeaders: nil,
+		PeerChanges:      nil,
+		RootHash:         []byte("root hash"),
+		MetaBlockHashes:  nil,
+		TxCount:          0,
+	}
+}
+
+//------- TestNewInterceptedHeader
```
Comment (sasurobert) [inline process/block/interceptedBlocks/interceptedBlockHeader_test.go]: delete comment.

File: process/block/interceptors/metachainHeaderInterceptor.go [hunk 4]
```diff
@@ -110,13 +112,13 @@ func (mhi *MetachainHeaderInterceptor) ProcessReceivedMessage(message p2p.Messag
 }
 
 func (mhi *MetachainHeaderInterceptor) processMetaHeader(metaHdrIntercepted *block.InterceptedMetaHeader) {
-	isHeaderOkForProcessing := mhi.headerValidator.IsHeaderValidForProcessing(metaHdrIntercepted.MetaBlock)
-	if !isHeaderOkForProcessing {
-		log.Debug("intercepted meta block header already processed")
+	err := mhi.headerValidator.HeaderValidForProcessing(metaHdrIntercepted)
+	if err != nil {
+		log.Debug(fmt.Sprintf("intercepted meta block header already processed: %s", err.Error()))
```
Comment (sasurobert) [inline process/block/interceptors/metachainHeaderInterceptor.go]: log.Debug("string" + err.Error()) no need for fmt.

File: process/block/interceptors/metachainHeaderInterceptor_test.go [hunk 5]
```diff
@@ -16,6 +15,8 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+var durTimeout = time.Second
```
Comment (sasurobert) [inline process/block/interceptors/metachainHeaderInterceptor_test.go]: timeoutDurration

File: process/factory/metachain/interceptorsContainerFactory.go [hunk 6]
```diff
@@ -11,19 +12,26 @@ import (
 	"github.com/ElrondNetwork/elrond-go/process/dataValidators"
 	"github.com/ElrondNetwork/elrond-go/process/factory"
 	"github.com/ElrondNetwork/elrond-go/process/factory/containers"
+	interceptors2 "github.com/ElrondNetwork/elrond-go/process/interceptors"
```
Comment (sasurobert) [inline process/factory/metachain/interceptorsContainerFactory.go]: change interceptors2

File: process/factory/metachain/interceptorsContainerFactory.go [hunk 7]
```diff
@@ -11,19 +12,26 @@ import (
...
+	interceptorFactory "github.com/ElrondNetwork/elrond-go/process/interceptors/factory"
+	"github.com/ElrondNetwork/elrond-go/process/interceptors/processor"
 	"github.com/ElrondNetwork/elrond-go/sharding"
 )
 
+const numGoRoutinesInMetaInterceptors = 100
```
Comment (sasurobert) [inline process/factory/metachain/interceptorsContainerFactory.go]: maxGoRoutines would be enough

File: process/interceptors/factory/argInterceptedDataFactory.go [hunk 8]
```diff
@@ -5,16 +5,25 @@ import (
...
+	*ArgMetaInterceptedDataFactory
+	KeyGen   crypto.KeyGenerator
+	Signer   crypto.SingleSigner
+	AddrConv state.AddressConverter
+}
+
+// ArgMetaInterceptedDataFactory holds all dependencies required by the meta intercepted data factory in order to create
+// new instances
+type ArgMetaInterceptedDataFactory struct {
+	Marshalizer         marshal.Marshalizer
```
Comment (sasurobert) [inline process/interceptors/factory/argInterceptedDataFactory.go]: when metachain receives a metachain, it does need to verify with single signer as well

File: process/block/interceptedBlocks/interceptedBlockHeader.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	)
+}
+
+// Hash gets the hash of this header
+func (inHdr *InterceptedHeader) Hash() []byte {
+	return inHdr.hash
+}
+
+// Shard returns the shard ID for which this header is addressed
+func (inHdr *InterceptedHeader) Shard() uint32 {
```
Comment (sasurobert) [inline process/block/interceptedBlocks/interceptedBlockHeader.go]: function not used

File: process/block/interceptedBlocks/interceptedBlockHeader.go (hunk 1 above)
Comment (iulianpascalau) [reply process/block/interceptedBlocks/interceptedBlockHeader.go]: So what should I do?

File: process/block/interceptedBlocks/interceptedBlockHeader_test.go (hunk 2 above)
Comment (iulianpascalau) [reply process/block/interceptedBlocks/interceptedBlockHeader_test.go]: Changed to dataBlock

File: process/block/interceptedBlocks/interceptedBlockHeader_test.go (hunk 3 above)
Comment (iulianpascalau) [reply process/block/interceptedBlocks/interceptedBlockHeader_test.go]: I would rather keep a limited number of placeholders only in test files as all tests start with Test... and is harder to keep the tests grouped by the tested func.

File: process/block/interceptors/metachainHeaderInterceptor.go (hunk 4 above)
Comment (iulianpascalau) [reply process/block/interceptors/metachainHeaderInterceptor.go]: changed

File: process/block/interceptors/metachainHeaderInterceptor_test.go (hunk 5 above)
Comment (iulianpascalau) [reply process/block/interceptors/metachainHeaderInterceptor_test.go]: changed

File: process/factory/metachain/interceptorsContainerFactory.go (hunk 6 above)
Comment (iulianpascalau) [reply process/factory/metachain/interceptorsContainerFactory.go]: changed to processInterceptors

File: process/factory/metachain/interceptorsContainerFactory.go (hunk 7 above)
Comment (iulianpascalau) [reply process/factory/metachain/interceptorsContainerFactory.go]: numGoRoutines shall be!

File: process/interceptors/factory/argInterceptedDataFactory.go (hunk 8 above)
Comment (iulianpascalau) [reply process/interceptors/factory/argInterceptedDataFactory.go]: Will be introduced whenever will be necessary. That is why I have implemented Arg... pattern as to easily add a new parameter.

File: process/block/interceptedBlocks/interceptedBlockHeader.go (hunk 9 above)
Comment (iulianpascalau) [reply process/block/interceptedBlocks/interceptedBlockHeader.go]: removed

File: process/factory/metachain/interceptorsContainerFactory.go [hunk 10]
```diff
@@ -63,16 +71,33 @@ func NewInterceptorsContainerFactory(
...
+		Marshalizer:         marshalizer,
+		Hasher:              hasher,
+		ShardCoordinator:    shardCoordinator,
+		ChronologyValidator: chronologyValidator,
+		MultiSigVerifier:    multiSigner,
+	}
+
+	icf := &interceptorsContainerFactory{
+		shardCoordinator:      shardCoordinator,
+		messenger:             messenger,
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply process/factory/metachain/interceptorsContainerFactory.go]: yes, changed to check.IfNil all checks

File: process/block/interceptedBlocks/interceptedBlockHeader.go (hunk 1 above)
Comment (sasurobert) [reply process/block/interceptedBlocks/interceptedBlockHeader.go]: can be set on constructor.

**Pull request merged**